torchao/prototype/quantized_training/int8_mixed_precision.py

```diff
@@ -0,0 +1,234 @@
+            else:
+                assert x.config == config
+            return x._data
+
+        args, kwargs = pytree.tree_map_only(cls, unwrap, (args, kwargs))
+        out = func(*args, **kwargs)
+        if func not in _ops_to_preserve_subclass:
+            return out
+        else:
+            return pytree.tree_map_only(Tensor, lambda x: cls(x, config), out)
+
+    def fsdp_pre_all_gather(self, mesh):
+        # TODO: pre-quantize weight here -> reduce comm bandwidth.
+        # we will need another tensor subclass to hold the quantized weight.
+        return (self._data,), (self.config,)
+
+    def fsdp_post_all_gather(
+        self,
+        all_gather_outputs: Tuple[Tensor, ...],
+        metadata: Any,
+        param_dtype: torch.dtype,
+        *,
+        out: Optional[Tensor] = None,
+    ):
+        (data,) = all_gather_outputs
+        (config,) = metadata
+        if out is not None:
+            assert isinstance(out, Int8MixedPrecisionLinearWeight)
+            assert out.config == config
+            return
+        return Int8MixedPrecisionLinearWeight(data.to(param_dtype), config), all_gather_outputs
+
+
+# alternative UX
+class Int8MixedPrecisionLinear(nn.Linear):
+    def __init__(self, *args, config: Int8MixedPrecisionConfig, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.config = config
+
+    def forward(self, input: Tensor) -> Tensor:
+        return _Int8MixedPrecisionLinear.apply(input, self.weight, self.bias, self.config)
+
+    def extra_repr(self):
+        return f"{super().extra_repr()}, config={self.config}"
+
+    @classmethod
+    def convert_linear(cls, module: nn.Module, config: Int8MixedPrecisionConfig = _DEFAULT_CONFIG):
+        if module.__class__ is nn.Linear:  # exact match, don't swap nn.Linear subclasses
+            module.__class__ = cls
+            module.config = config
+            return
+        for child in module.children():
+            cls.convert_linear(child, config)
+
+
+def _dynamic_int8_mm(A: Tensor, B: Tensor) -> Tensor:
+    # INT8 matmul is the most performant when A is row-major and B is column-major.
+    # thus, we transpose B before quantization.
+
+    # A and B (before quantization) might not be contiguous, especially in backward.
+    # it's also not guaranteed that A_i8 and B_t_i8 (after quantization) are contiguous,
+    # thus we have to call .contiguous() on them.
+    # hope that the .contiguous() calls will be fused into quantize op by torch.compile()
+    # TODO: investigate if calling .contiguous() before quantization is better.
+    # TODO: check if transpose+quantize are fused.
+    A_i8, A_scale_rowwise = quantize_int8_rowwise(A)
+    B_t_i8, B_scale_colwise = quantize_int8_rowwise(B.T)
+    return int8_mm_dequant(
+        A_i8.contiguous(),
+        B_t_i8.contiguous().T,
+        A_scale_rowwise,
+        B_scale_colwise,
+    )
+
+
+class _Int8MixedPrecisionLinear(torch.autograd.Function):
```

Comment:
@vkuzo I notice I might need to add `@torch._dynamo.allow_in_graph` here like what is done in FP8 for certain things to work. Do you have any idea why it is the case?

Also, it seems like calling a custom autograd.Function under `__torch_function__` (NOT `__torch_dispatch__`) results in silent bugs, even in eager mode. (I know that things under `__torch_dispatch__` doesn't have gradients/autograd is not working at dispatch level). Do you know if "calling a custom autograd.Function under `__torch_function__`" is supported in general?

Reply:
> I notice I might need to add @torch._dynamo.allow_in_graph here like what is done in FP8 for certain things to work. Do you have any idea why it is the case?

that decorator allows the "create the subclass" node to live in one level of the torch.compile IR without being decomposed (the decomposition happens later).  At the time when float8 was built, the features we needed for float8 only worked with this decorator, and not when the constructor was decomposed right away.  I hear this might be different now, but I haven't checked myself (because the current way works fine).